nemo/collections/llm/gpt/data/megatron/hyena/evo2_dataset.py

```diff
@@ -217,35 +219,52 @@ def process_segment(seg_seq: torch.Tensor) -> torch.Tensor:
+                # fastest check is to look at the first token after the pipe, if it is a '*_' then the
                 # tag starts _after_ the pipe, otherwise it starts before.
-                next_tok = seg_seq[first_pipe + 1].item()
-                if next_tok == first_taxonomy_prefix_token:
-                    # 'd' character for domain, which is the first part of a phylo tag.
-                    # tag starts after the pipe.
+                if len(seg_seq) > first_pipe + 2:
+                    first_tok = seg_seq[first_pipe + 1].item()
+                    next_tok = seg_seq[first_pipe + 2].item()
+                    if next_tok == 95 or first_tok == 59:  # ord('_') = 95, ord(';') = 59
+                        # the syntax is [tag char]_ or a missing segment would start with ;
+                        is_tag = False
+                    else:
+                        # tag starts before the pipe.
+                        is_tag = True
+                elif len(seg_seq) > first_pipe + 1:
+                    next_tok = seg_seq[first_pipe + 1].item()
+                    # Much weaker signal since these are sometimes degenerate bases, but at least the length
+                    #  check later will prevent overmasking.
+                    if next_tok in {68, 100, 82, 114, 59}:  # D,d,R,r for domain or realm (viruses) or missing ;
+                        is_tag = False
+                    else:
+                        is_tag = True
+                elif first_pipe >= max_tag_len or region_all_valid_or_control(seg_seq[:first_pipe]):
                     is_tag = False
                 else:
-                    # tag starts before the pipe.
                     is_tag = True
             else:
                 # The sequence ends with a pipe, so just check everything before the pipe and return the seg mask
                 assert first_pipe == seg_len - 1
                 # The sequence ends with a pipe, so just check everything before the pipe.
-                if region_all_valid_or_control(seg_seq[:first_pipe]):
+                if first_pipe >= max_tag_len or region_all_valid_or_control(seg_seq[:first_pipe]):
                     return seg_mask  # Pipe pos has already been masked
                 else:
                     seg_mask[:first_pipe] = 0
                     return seg_mask
             start = 0
             for end in pipe_pos:
-                if is_tag:
+                seg_len = end - start
+                if is_tag and seg_len < max_tag_len:  # Prefer faulty tag mask logic over masking entire seqs
                     seg_mask[start:end] = 0
+                elif is_tag and seg_len >= max_tag_len:
+                    is_tag = False  # Flip the state machine because the is_tag=True was for a very long region
                 else:
                     pass
                 is_tag = not is_tag  # Flip the state machine.
```

Comment:
So, if I understand this correctly. on line 259 we say "If we have a tag but it's too long, we'll just not label it as a tag anymore":

```
elif is_tag and seg_len >= max_tag_len: is_tag = False
```

But then, here on line 263, after the conditionals run, we flip things back? `is_tag - not_is_tag`?

Reply:
Addressed with comments inline.